### PR TITLE
fix: loss computation after embeddings resize - mllama

### PR DESCRIPTION
**Pull request by Ssukriti**
# What does this PR do?

background: https://github.com/huggingface/transformers/pull/36591

Fixes loss computation when vocab is resized by resize_embeddings. Only vocab size of parent conditionalgenerationclass is modified, hence loss has to be calculated there. Chosen approach until bigger refactor. Solution discussed with @zucchini-nlp in above PR 

Fixes # (issue)
https://github.com/huggingface/transformers/issues/36590

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@zucchini-nlp as discussed solution

### CODE REVIEW

File: src/transformers/models/mllama/modeling_mllama.py [hunk 1]
```diff
@@ -1972,16 +1970,11 @@ def forward(
...
-        if labels is not None:
-            loss = self.loss_function(logits, labels, self.vocab_size, **loss_kwargs)
-
         if not return_dict:
             output = (logits,) + outputs[1:]
-            return (loss,) + output if loss is not None else output
+            return output
 
         return CausalLMOutputWithPast(
-            loss=loss,
```
Comment (Ssukriti) [inline src/transformers/models/mllama/modeling_mllama.py:1984]: I still left rest of return types in the class , for minimal changes as you mentioned there would be a massive refactor @zucchini-nlp . So I just moved loss computation to ConditionalGenerationClass . However with that change, there may not be a need for this `MllamaForCausalLM` as basically its just adding logits, which can also be moved and just the MllamaText Model class can be used directly from wrapper.

Will leave that to you as you think of the refactor. Or if you want me to clean it up, can do

File: src/transformers/models/mllama/modeling_mllama.py [hunk 2]
```diff
@@ -1902,14 +1902,12 @@ def forward(
         full_text_row_masked_out_mask: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,
         past_key_values: Optional[Union[Cache, List[torch.FloatTensor]]] = None,
         inputs_embeds: Optional[torch.FloatTensor] = None,
-        labels: Optional[torch.LongTensor] = None,
```
Comment (Ssukriti) [inline src/transformers/models/mllama/modeling_mllama.py:1905]: in order for common tests to pass, I may have to keep this and leave it unused. the common tests pass labels to forward for that CausalLM class mapping types , so it results in invalid argument otherwise. Let me know if any better ideas for now that are aligned with any refactors planned

File: tests/models/mllama/test_modeling_mllama.py [hunk 3]
```diff
@@ -321,6 +321,27 @@ def test_inputs_embeds_matches_input_ids(self):
                 out_embeds = model(inputs_embeds=inputs_embeds, **inputs)[0]
             torch.testing.assert_close(out_embeds, out_ids)
 
+    def test_resize_embeddings_results_in_successful_loss(self):
```
Comment (Ssukriti) [inline tests/models/mllama/test_modeling_mllama.py:324]: test for reported bug that would fail earlier and now pass

**Review by zucchini-nlp — COMMENTED** [review]
Great, thanks a lot! This looks better to me as a temporary workaround until refactoring. I left some comments about removing `loss` from causalLM completely, but otherwise LGTM

File: src/transformers/models/mllama/modeling_mllama.py (hunk 1 above)
Comment (Ssukriti) [reply src/transformers/models/mllama/modeling_mllama.py:1984]: I will add it back for users that use this class directly, but will not pass labels from GenerationClass to avoid exception

File: tests/models/mllama/test_modeling_mllama.py [hunk 4]
```diff
@@ -321,6 +321,27 @@ def test_inputs_embeds_matches_input_ids(self):
...
+        # resizing embeddings should result in successful loss computation
+        config, inputs = self.model_tester.prepare_config_and_inputs_for_common()
+        labels = torch.zeros(
+                    (self.model_tester.batch_size, self.model_tester.seq_length), dtype=torch.long, device=torch_device
+                )
```
[the remark this answers is not in the document]
Comment (Ssukriti) [reply tests/models/mllama/test_modeling_mllama.py]: thank you :), I missed that. addressed

File: tests/models/mllama/test_modeling_mllama.py [hunk 5]
```diff
@@ -321,6 +321,27 @@ def test_inputs_embeds_matches_input_ids(self):
...
+
+        for model_class in self.all_model_classes:
+            model = model_class(config)
+            model_vocab_size = config.get_text_config().vocab_size
+            # Resize embeddings and call forward
+            model.resize_token_embeddings(model_vocab_size + 10)
+            with torch.autocast(device_type="cuda", dtype=torch.float16):
```
[the remark this answers is not in the document]
Comment (Ssukriti) [reply tests/models/mllama/test_modeling_mllama.py]: sounds good

**Comment by Ssukriti** [comment]
@zucchini-nlp all tests have passed and comments addressed . Thank you for the review.

**Review by zucchini-nlp — APPROVED** [review]
Perfect, thanks!

**Comment by Ssukriti** [comment]
Thank you @zucchini-nlp . Can the PR be merged soon as well ? It is actually blocking a use case we have

**Comment by zucchini-nlp** [comment]
Yep, merging, sorry
